File: Best Time to Buy and Sell Stock II.py
def maxProfit(prices):
    maxProfit = 0
    # Start from the second day
    for i in range(1, len(prices)):
        # If the price of the stock has increased compared to the previous day
        if prices[i] > prices[i - 1]:
            # Add the difference to maxProfit
            maxProfit += prices[i] - prices[i - 1]
    # Return the total profit
    return maxProfit

# Example usage:
prices1 = [7, 1, 5, 3, 6, 4]
prices2 = [1, 2, 3, 4, 5]
prices3 = [7, 6, 4, 3, 1]

print(maxProfit(prices1)) # Output: 7
print(maxProfit(prices2)) # Output: 4
print(maxProfit(prices3)) # Output: 0



# SOLUTION 2ND 

def maxProfit(prices):
    return sum(max(prices[i+1] - prices[i], 0) for i in range(len(prices)-1))

# Example usage:
prices1 = [7, 1, 5, 3, 6, 4]
prices2 = [1, 2, 3, 4, 5]
prices3 = [7, 6, 4, 3, 1]

print(maxProfit(prices1)) # Output: 7
print(maxProfit(prices2)) # Output: 4
print(maxProfit(prices3)) # Output: 0

File: Best Time to Buy and Sell Stock II.js
function maxProfit(prices) {
    let maxProfit = 0;
    // Start from the second day
    for (let i = 1; i < prices.length; i++) {
        // If the price of the stock has increased compared to the previous day
        if (prices[i] > prices[i - 1]) {
            // Add the difference to maxProfit
            maxProfit += prices[i] - prices[i - 1];
        }
    }
    // Return the total profit
    return maxProfit;
}

// Example usage:
const prices1 = [7, 1, 5, 3, 6, 4];
const prices2 = [1, 2, 3, 4, 5];
const prices3 = [7, 6, 4, 3, 1];

console.log(maxProfit(prices1)); // Output: 7
console.log(maxProfit(prices2)); // Output: 4
console.log(maxProfit(prices3)); // Output: 0


// SOLUTION 2ND

const maxProfit = prices => 
  prices.reduce((profit, price, i, arr) => 
    profit + (i > 0 ? Math.max(0, price - arr[i - 1]) : 0), 0);

// Example usage:
const prices1 = [7, 1, 5, 3, 6, 4];
const prices2 = [1, 2, 3, 4, 5];
const prices3 = [7, 6, 4, 3, 1];

console.log(maxProfit(prices1)); // Output: 7
console.log(maxProfit(prices2)); // Output: 4
console.log(maxProfit(prices3)); // Output: 0
